### ISSUE TITLE
Java 9+ with module system: could not locate stub library in jar file

### ISSUE THREAD

**Issue by oobles** [jnr/jffi#73]
I'm converting a project to use Java 11 using the module system.  Using AdoptOpenJdk 11.0.4.  Getting the following error.

Caused by: java.lang.UnsatisfiedLinkError: java.lang.UnsatisfiedLinkError: could not locate stub library in jar file.  Tried [jni/Darwin/libjffi-1.2.dylib, /jni/Darwin/libjffi-1.2.dylib]
	at jffi@1.2.19/com.kenai.jffi.internal.StubLoader.getStubLibraryStream(StubLoader.java:416)
	at jffi@1.2.19/com.kenai.jffi.internal.StubLoader.loadFromJar(StubLoader.java:359)
	at jffi@1.2.19/com.kenai.jffi.internal.StubLoader.load(StubLoader.java:262)
	at jffi@1.2.19/com.kenai.jffi.internal.StubLoader.<clinit>(StubLoader.java:453)

This seems to be related to how Java 9+ deals with getResourceAsStream. It doesn't allow accessing another jar's resources, so the files in jffi-1.2.19-native.jar are not resolved.

https://docs.oracle.com/javase/9/docs/api/java/lang/Class.html#getResourceAsStream-java.lang.String-

Using some jar command line hacks I was able to combine the contents of jffi-1.2.19-native.jar into the jffi-1.2.19.jar.  As the resource is now in the same jar file the getResourceAsStream successfully finds the native library.

mkdir contents
cd contents
jar -xf ../jffi-1.2.19-native.jar
jar -xf ../jffi-1.2.19.jar
jar -cfm ../jffi-1.2.19.jar META-INF/MANIFEST.MF .

The other option would be to move the StubLoader.class or similar class into the jffi-1.2.19-native.jar and then try using that class when loading.

**Comment by V-F** [comment]
Hello, @headius.
Is supporting of modular architecture planned?
I wanted to ask to add the `Automatic-Module-Name` to the `MANIFEST.MF` file, since in a modular application the both files (eg. jffi-1.2.23.jar and jffi-1.2.23-native.jar) have the same module name:
```Error occurred during initialization of boot layer
java.lang.module.FindException: Two versions of module jffi found in X:\release (jffi-1.2.23.jar and jffi-1.2.23-native.jar)
```
But considering the above problem the only solution would be also to combine the content of both files into the new one.

**Comment by headius** [comment]
@V-F Is this not already fixed by 4b198d2798cf9f2ba65e8a17485596d33c0a4106? That change went out in 1.3.1.

If there's something more we need to do, please let us know!

**Comment by headius** [comment]
FWIW if this is still an issue I do not know of a reason we have to keep separating out the native jar. We do not release them independently.

However it seems like this is a more general problem. My understanding was that this `native` artifact is the proper way to package a JNI library with Maven, in which case how do other projects then access that library given module restrictions?

We are happy to do the right thing here, but what is the right thing?

**Comment by V-F** [comment]
Thanks for the quick response.
You are right, I didn't notice that the manifest in the jffi-1.3.1-native.jar file already contains `Automatic-Module-Name`, which allows to distinguish this automatic module from the jffi automatoc module of the jffi-1.3.1.jar file (which doesn't contain `Automatic-Module-Name` in it manifest) and this solves my problem.
Regarding the error indicated by @oobles at the moment I have no way to check for its existence, but considering that the resource is not [encapsulated](https://docs.oracle.com/javase/9/docs/api/java/lang/Module.html#getResourceAsStream-java.lang.String-) and the automatic module exports all it packages, everything should work.
It would be great if all jnr-libraries had a module descriptor, then they could be used by jlink without additional effort.

**Comment by headius** [comment]
> this solves my problem

Great! Thank you for confirming.

> It would be great if all jnr-libraries had a module descriptor

Are you looking for an explicit module-info rather than automatic module name? All the JNR libraries gained automatic module names plus build and code fixes to support Java 9+ a few months back. Adding module-info wouldn't be difficult at this point, but it would force the build to always use Java 9+ (currently it can still build on Java 8).

**Comment by headius** [comment]
> force the build to always use Java 9+

... or at least make the build more complicated to keep supporting Java 8. It is a bit cumbersome to provide module-info in a jar that still works on Java 8, which is why we only did automatic module names.

**Comment by headius** [comment]
I am going to close this as fixed in 1.3.1, since that is the release when we gave the native jar its own module name.

@oobles @V-F If there's further work needed, like fully going to module-info, please open a new issue for that (or a PR ❤️).
